src/NHibernate/Action/CollectionRecreateAction.cs

```diff
@@ -30,9 +30,12 @@ public override void Execute()
 
 			PreRecreate();
 
-			Persister.Recreate(collection, Key, Session);
+			var key = GetKey();
+			Persister.Recreate(collection, key, Session);
 
-			Session.PersistenceContext.GetCollectionEntry(collection).AfterAction(collection);
+			var entry = Session.PersistenceContext.GetCollectionEntry(collection);
+			entry.CurrentKey = key;
```

Comment:
I think setting `CurrentKey` here is more correct than setting only `loadedKey` in `AfterAction`.

Also, this seems the only place where the current key has to be updated.